test/dependency_services/dependency_services_test.dart

```diff
@@ -0,0 +1,212 @@
+
+    return '''
+\$ cat $filename
+${contents.join('\n')}''';
+  }
+
+  context.expectNextSection('''
+${catFile('pubspec.yaml')}
+${catFile('pubspec.lock')}
+''');
+}
+
+late final String snapshot;
+
+extension on GoldenTestContext {
+  Future<void> runDependencyServices(List<String> args, {String? stdin}) async {
+    final buffer = StringBuffer();
+    buffer.writeln('## Section ${args.join(' ')}');
+    final process = await Process.start(
+      Platform.resolvedExecutable,
+      [
+        snapshot,
+        ...args,
+      ],
+      environment: getPubTestEnvironment(),
+      workingDirectory: p.join(d.sandbox, appPath),
+    );
+    if (stdin != null) {
+      process.stdin.write(stdin);
+      await process.stdin.flush();
+      await process.stdin.close();
+    }
+    final outLines = outputLines(process.stdout);
+    final errLines = outputLines(process.stderr);
+    final exitCode = await process.exitCode;
+
+    final pipe = stdin == null ? '' : ' echo ${protectArgument(stdin)} |';
+    buffer.writeln([
+      '\$$pipe dependency_services ${args.map(protectArgument).join(' ')}',
+      ...await outLines,
+      ...(await errLines).map((e) => '[STDERR] $e'),
+      if (exitCode != 0) '[EXIT CODE] $exitCode',
+    ].join('\n'));
+
+    expectNextSection(buffer.toString());
+  }
+}
+
+Future<Iterable<String>> outputLines(Stream<List<int>> stream) async {
+  final s = await utf8.decodeStream(stream);
+  if (s.isEmpty) return [];
+  return filterUnstableLines(s.split('\n'));
+}
+
+Future<void> listReportApply(
+  GoldenTestContext context,
+  List<_PackageVersion> upgrades,
+) async {
+  manifestAndLockfile(context);
+  await context.runDependencyServices(['list']);
+  await context.runDependencyServices(['report']);
+
+  final input = json.encode({
+    'dependencyChanges': upgrades,
+  });
+
+  await context.runDependencyServices(['apply'], stdin: input);
+  manifestAndLockfile(context);
+}
+
+Future<void> main() async {
+  setUpAll(() async {
+    final tempDir = Directory.systemTemp.createTempSync();
+    snapshot = p.join(tempDir.path, 'dependency_services.dart.snapshot');
+    final r = Process.runSync(Platform.resolvedExecutable, [
+      '--snapshot=$snapshot',
+      p.join('bin', 'dependency_services.dart'),
+    ]);
+    expect(r.exitCode, 0, reason: r.stderr);
+  });
+
+  tearDownAll(() {
+    File(snapshot).parent.deleteSync(recursive: true);
+  });
+
+  testWithGolden('Removing transitive', (context) async {
+    (await servePackages())
+      ..serve('foo', '1.2.3', deps: {'transitive': '^1.0.0'})
+      ..serve('foo', '2.2.3')
+      ..serve('transitive', '1.0.0');
+
+    await d.dir(appPath, [
+      d.pubspec({
+        'name': 'app',
+        'dependencies': {
+          'foo': '^1.0.0',
+        },
+      })
+    ]).create();
+    await pubGet();
+    await listReportApply(context, [
+      _PackageVersion('foo', Version.parse('2.2.3')),
+      _PackageVersion('transitive', null)
+    ]);
```

Comment:
As much as I love golden tests, maybe we should add some assertions here?

There is a non-zero risk that we won't notice minor changes in the golden files.

If we could do a simple assertion like:
```dart
expect(report['dependencies'].any((depReport) => depReport['name'] == 'foo' && depReport['singleBreaking'].any(
  (d) => d['name'] == 'transitive' && d['version'] == null),
), isTrue);
```

Or something like that... just a few, we don't have to check every possible property :D